File: app/src/main/java/ca/judacribz/gainzassist/models/db/WorkoutViewModel.java
package ca.judacribz.gainzassist.models.db;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.content.Context;
import ca.judacribz.gainzassist.models.Exercise;
import ca.judacribz.gainzassist.models.ExerciseSet;
import ca.judacribz.gainzassist.models.Session;
import ca.judacribz.gainzassist.models.Workout;

import java.util.List;

public class WorkoutViewModel extends AndroidViewModel {
    private WorkoutRepo workoutRepo;

    public WorkoutViewModel(Application app) {
        super(app);
        workoutRepo = new WorkoutRepo(app);
    }

    // CREATE
    // --------------------------------------------------------------------------------------------
    public void insertWorkout(Workout... workouts) {
        for (Workout workout : workouts)
            workoutRepo.insertWorkout(workout);
    }

    public void insertExercise(Exercise... exercises) {
        for (Exercise exercise : exercises)
            workoutRepo.insertExercise(exercise);
    }

    public void insertSession(Session... sessions) {
        for (Session session : sessions)
            workoutRepo.insertSession(session, true);
    }

    public void insertSet(ExerciseSet... exerciseSets) {
        for (ExerciseSet exerciseSet : exerciseSets)
            workoutRepo.insertSet(exerciseSet);
    }
    // --------------------------------------------------------------------------------------------


    // RETRIEVE
    // --------------------------------------------------------------------------------------------
    public LiveData<List<Workout>> getAllWorkouts() {
        return workoutRepo.getAllWorkoutsLive();
    }
    public LiveData<List<Session>> getAllSessions() {
        return workoutRepo.getAllSessionsLive();
    }
    public LiveData<List<ExerciseSet>> getAllSets() {
        return workoutRepo.getAllSetsLive();
    }

    public LiveData<Workout> getWorkout(int id) {
        return workoutRepo.getWorkout(id);
    }

    public void getWorkoutFromName(Context context, String name) {
        workoutRepo.getWorkoutFromName(context, name);
    }

    public LiveData<List<Exercise>> getExercisesFromWorkout(int workoutId) {
        return workoutRepo.getExercisesFromWorkout(workoutId);
    }

    public LiveData<Exercise> getExercise(long id) {
        return workoutRepo.getExercise(id);
    }

    public LiveData<List<String>> getAllUniqueExerciseNames() {
        return workoutRepo.getAllUniqueExerciseNames();
    }

    public LiveData<List<ExerciseSet>> getSetsFromExercise(long exerciseId) {
        return workoutRepo.getSetsFromExercise(exerciseId);
    }

    // --------------------------------------------------------------------------------------------


    // UPDATE
    // --------------------------------------------------------------------------------------------
    void updateWorkout(Workout workout) {
        workoutRepo.updateWorkout(workout);
    }

    void updateExercise(Exercise exercise) {
        workoutRepo.updateExercise(exercise);
    }

    void updateSet(ExerciseSet exerciseSet) {
        workoutRepo.updateSet(exerciseSet);
    }
    // --------------------------------------------------------------------------------------------


    // DELETE
    // --------------------------------------------------------------------------------------------
    public void deleteAllWorkouts() {
        workoutRepo.deleteAllWorkouts();
    }

    public void deleteWorkout(Workout workout) {
        workoutRepo.deleteWorkout(workout);
    }

    public void deleteExercise(Exercise exercise) {
        workoutRepo.deleteExercise(exercise);
    }

    public void deleteSet(ExerciseSet exerciseSet) {
        workoutRepo.deleteSet(exerciseSet);
    }
    // --------------------------------------------------------------------------------------------
}

File: app/src/main/java/ca/gainzassist/models/db/WorkoutViewModel.kt
package ca.gainzassist.models.db

import android.app.Application
import android.content.Context
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import ca.gainzassist.models.Exercise
import ca.gainzassist.models.ExerciseSet
import ca.gainzassist.models.Session
import ca.gainzassist.models.Workout

class WorkoutViewModel(app: Application) : AndroidViewModel(app) {
    private val workoutRepo = WorkoutRepo(app)

    // CREATE
    // --------------------------------------------------------------------------------------------
    fun insertWorkout(vararg workouts: Workout) {
        for (workout in workouts) {
            workoutRepo.insertWorkout(workout)
        }
    }

    fun insertExercise(vararg exercises: Exercise) {
        for (exercise in exercises) {
            workoutRepo.insertExercise(exercise)
        }
    }

    fun insertSession(vararg sessions: Session) {
        for (session in sessions) {
            workoutRepo.insertSession(session, true)
        }
    }

    fun insertSet(vararg exerciseSets: ExerciseSet) {
        for (exerciseSet in exerciseSets) {
            WorkoutRepo.insertSet(exerciseSet)
        }
    }
    // --------------------------------------------------------------------------------------------

    // RETRIEVE
    // --------------------------------------------------------------------------------------------
    val allWorkouts: LiveData<List<Workout>>
        get() = workoutRepo.getAllWorkoutsLive()

    val allSessions: LiveData<List<Session>>
        get() = workoutRepo.getAllSessionsLive()

    val allSets: LiveData<List<ExerciseSet>>
        get() = workoutRepo.getAllSetsLive()

    fun getWorkout(id: Int): LiveData<Workout> {
        return workoutRepo.getWorkout(id.toLong())
    }

    fun getWorkoutFromName(context: Context?, name: String) {
        workoutRepo.getWorkoutFromName(context, name)
    }

    fun getExercisesFromWorkout(workoutId: Int): LiveData<List<Exercise>> {
        return workoutRepo.getExercisesFromWorkout(workoutId.toLong())
    }

    fun getExercise(id: Long): LiveData<Exercise> {
        return workoutRepo.getExercise(id)
    }

    val allUniqueExerciseNames: LiveData<List<String>>
        get() = workoutRepo.getAllUniqueExerciseNames()

    fun getSetsFromExercise(exerciseId: Long): LiveData<List<ExerciseSet>> {
        return workoutRepo.getSetsFromExercise(exerciseId)
    }
    // --------------------------------------------------------------------------------------------

    // UPDATE
    // --------------------------------------------------------------------------------------------
    fun updateWorkout(workout: Workout) {
        workoutRepo.updateWorkout(workout)
    }

    internal fun updateExercise(exercise: Exercise) {
        workoutRepo.updateExercise(exercise)
    }
    // --------------------------------------------------------------------------------------------

    // DELETE
    // --------------------------------------------------------------------------------------------
    fun deleteAllWorkouts() {
        workoutRepo.deleteAllWorkouts()
    }

    fun deleteWorkout(workout: Workout) {
        workoutRepo.deleteWorkout(workout)
    }

    fun deleteWorkout(workoutName: String) {
        workoutRepo.deleteWorkout(workoutName)
    }

    fun deleteExercise(exercise: Exercise) {
        workoutRepo.deleteExercise(exercise)
    }

    fun deleteSet(exerciseSet: ExerciseSet) {
        workoutRepo.deleteSet(exerciseSet)
    }
    // --------------------------------------------------------------------------------------------
}
